### PR TITLE
Migrate dealer wallet to use graphql api

### PR DESCRIPTION
**Pull request by sebastienverreault**
Migrating dealer from inherited wallet to graphql wallet
First implementation will mock the few required endpoints until galoy api v2 is finalized

### CODE REVIEW

File: src/Dealer.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        withdrawOnChainAddress,
+        this.withdrawBookKeeping,
+        this.depositOnExchangeCallback,
+      )
+      result.updatedLeverageResult = updatedLeverageResult
+      if (updatedLeverageResult.ok) {
+        const updatedLeverage = updatedLeverageResult.value
+        logger.info(
+          { updatedLeverage },
+          `The active ${activeStrategy} strategy was successful at UpdateLeverage()`,
```
Comment (nicolasburtey) [inline src/Dealer.ts:191]: `activeStrategy` should be passed as an argument in the first obj instead of the description.

File: src/Dealer.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
+import pino from "pino"
```
Comment (nicolasburtey) [inline src/Dealer.ts:1]: filename `Dealer.ts` is probably too generic. server.ts or entrypoint.ts maybe?

File: src/DealerMockWallet.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  public async getWalletOnChainDepositAddress(): Promise<Result<string>> {
+    const logger = this.logger.child({ method: "getWalletOnChainDepositAddress()" })
+    try {
+      const result = await this.client.query({ query: GET_ONCHAIN_ADDRESS })
+      logger.debug({ result })
+      return { ok: true, value: result.data.getLastOnChainAddress.id }
+    } catch (error) {
+      logger.error({ error })
```
Comment (nicolasburtey) [inline src/DealerMockWallet.ts:97]: a message is needed here in the second argument

File: src/DealerMockWallet.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      return { ok: false, error }
+    }
+  }
+
+  public async getWalletOnChainDepositAddress(): Promise<Result<string>> {
+    const logger = this.logger.child({ method: "getWalletOnChainDepositAddress()" })
+    try {
+      const result = await this.client.query({ query: GET_ONCHAIN_ADDRESS })
+      logger.debug({ result })
```
Comment (nicolasburtey) [inline src/DealerMockWallet.ts:94]: a message is needed here in the second argument

File: src/DealerMockWallet.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    }
+  }
+`
+
+export class DealerMockWallet implements GaloyWallet {
+  client: ApolloClient<NormalizedCacheObject>
+  logger: pino.Logger
+
+  constructor(logger: pino.Logger) {
+    const GRAPHQL_URI = process.env["GRAPHQL_URI"]
```
Comment (nicolasburtey) [inline src/DealerMockWallet.ts:71]: this should be passed as an argument to the constructor.

File: src/DealerMockWallet.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    const cache = new InMemoryCache(IN_MEMORY_CACHE_CONFIG)
+    this.client = new ApolloClient({ link: httpLink, cache: cache })
+    this.logger = logger.child({ class: DealerMockWallet.name })
+  }
+
+  public async getWalletUsdBalance(): Promise<Result<number>> {
+    const logger = this.logger.child({ method: "getWalletUsdBalance()" })
+    try {
+      const result = await this.client.query({ query: WALLET })
+      logger.debug({ result })
```
Comment (nicolasburtey) [inline src/DealerMockWallet.ts:82]: a message is needed here in the second argument

File: src/DealerMockWallet.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      return { ok: true, value: result.data.wallet.balance.amount }
+    } catch (error) {
+      logger.error({ error })
```
Comment (nicolasburtey) [inline src/DealerMockWallet.ts:85]: a message is needed here in the second argument

File: src/ExchangeBase.ts [hunk 8]
```diff
@@ -66,6 +68,62 @@ export abstract class ExchangeBase {
     }
   }
 
+  public async fetchDeposits(
+    address: string,
```
Comment (nicolasburtey) [inline src/ExchangeBase.ts:72]: you should use object input instead of positional argument

Comment (nicolasburtey) [reply src/ExchangeBase.ts:72]: ie: `{address, amountInSats}`

File: src/ExchangeBase.ts [hunk 9]
```diff
@@ -66,6 +68,62 @@ export abstract class ExchangeBase {
...
+    amountInSats: number,
+  ): Promise<Result<FetchDepositsResult>> {
+    try {
+      this.exchangeConfig.fetchDepositsValidateInput(address, amountInSats)
+
+      // No way to filter on request, so get them all...
+      const response = await this.fetchDepositsAllPages()
+      this.logger.debug(
+        { response },
+        `exchange.fetchDeposits(${address}, ${amountInSats}) returned: {response}`,
```
Comment (nicolasburtey) [inline src/ExchangeBase.ts:82]: move variables to the first argument

File: src/ExchangeBase.ts [hunk 10]
```diff
@@ -66,6 +68,62 @@ export abstract class ExchangeBase {
...
+      )
+
+      // ...then filter
+      const result = this.exchangeConfig.fetchDepositsProcessApiResponse(
+        address,
```
Comment (nicolasburtey) [inline src/ExchangeBase.ts:87]: use object instead of positional arguments

File: src/ExchangeBase.ts [hunk 11]
```diff
@@ -66,6 +68,62 @@ export abstract class ExchangeBase {
...
+      const code = undefined
+      const since = undefined
+      const limit = 100
+      const params = {
+        page: page,
+      }
+      const deposits = await this.exchange.fetchDeposits(code, since, limit, params)
+      if (deposits.length) {
+        page = this.exchange.last_json_response["cursor"]
+        console.log(`Got fetchDeposits() page=${page}`)
```
Comment (nicolasburtey) [inline src/ExchangeBase.ts:115]: be consistent on the logging used

File: src/FtxExchangeConfiguration.ts [hunk 12]
```diff
@@ -44,6 +47,45 @@ export class FtxExchangeConfiguration implements ExchangeConfiguration {
     }
   }
 
+  fetchDepositsValidateInput(address: string, amountInSats: number) {
```
Comment (nicolasburtey) [inline src/FtxExchangeConfiguration.ts:50]: what is the goal of this function?

File: src/FtxExchangeConfiguration.ts [hunk 13]
```diff
@@ -44,6 +47,45 @@ export class FtxExchangeConfiguration implements ExchangeConfiguration {
...
+    assert(address, ApiError.UNSUPPORTED_ADDRESS)
+    assert(amountInSats, ApiError.NON_POSITIVE_QUANTITY)
+    assert(amountInSats > 0, ApiError.NON_POSITIVE_QUANTITY)
+  }
```
Comment (nicolasburtey) [inline src/FtxExchangeConfiguration.ts:54]: did you enable the linter? there is no space between the 2 functions for some reasons

File: src/FtxExchangeConfiguration.ts [hunk 14]
```diff
@@ -44,6 +47,45 @@ export class FtxExchangeConfiguration implements ExchangeConfiguration {
...
+  fetchDepositsProcessApiResponse(
```
Comment (nicolasburtey) [inline src/FtxExchangeConfiguration.ts:55]: use object instead of positional arguments

File: src/FtxExchangeConfiguration.ts [hunk 15]
```diff
@@ -44,6 +47,45 @@ export class FtxExchangeConfiguration implements ExchangeConfiguration {
...
+    address: string,
+    amountInSats: number,
+    response,
+  ): FetchDepositsResult {
+    assert(response, ApiError.UNSUPPORTED_API_RESPONSE)
+    const amountInBtc = sat2btc(amountInSats)
+    // We should receive an array of array of response objects
+    // Since we're looking for one specific entry, don't fail until we're done
+    const result = {} as FetchDepositsResult
+    response.forEach((responses) => {
```
Comment (nicolasburtey) [inline src/FtxExchangeConfiguration.ts:65]: response and responses are probably inverted here

**Review by nicolasburtey — CHANGES_REQUESTED** [review]
some early feedbacks that needs to be addressed. need to dive deeper into the PR later this week.

File: src/Dealer.ts (hunk 1 above)
Comment (sebastienverreault) [reply src/Dealer.ts:191]: done, updated a bunch of them

File: src/DealerMockWallet.ts (hunk 3 above)
Comment (sebastienverreault) [reply src/DealerMockWallet.ts:97]: Added messages to all logs of that file

File: src/ExchangeBase.ts (hunk 8 above)
Comment (sebastienverreault) [reply src/ExchangeBase.ts:72]: done

File: src/ExchangeBase.ts (hunk 9 above)
Comment (sebastienverreault) [reply src/ExchangeBase.ts:82]: done

File: src/ExchangeBase.ts (hunk 11 above)
Comment (sebastienverreault) [reply src/ExchangeBase.ts:115]: that did not belong there, removed it

File: src/FtxExchangeConfiguration.ts (hunk 13 above)
Comment (sebastienverreault) [reply src/FtxExchangeConfiguration.ts:54]: yes I did and run it before commits, it seems to accept 0 or 1, but removes extra if any.

File: src/FtxExchangeConfiguration.ts (hunk 15 above)
Comment (sebastienverreault) [reply src/FtxExchangeConfiguration.ts:65]: poor naming, updated with deposits

**Pull request merged**